File: leetcode/c++/13.cpp
class Solution {
public:
    int romanToInt(string s) {
        int bases[] = {1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        string romans[] = {"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        int res = 0;
        for (int i = 0; i < 13; i++) {
            int base = bases[i];
            string roman = romans[i];
            while (s.size() >= roman.size()) {
                if (s.substr(0, roman.size()) == roman) {
                    res += base;
                    s = s.substr(roman.size());
                } else {
                    break;
                }
            }
        }
        return res;
    }
};

File: leetcode/python/13.py
#!/usr/bin/env python

__author__ = "yangyanzhan"
__email__ = "yangyanzhan@gmail.com"
__url__ = "https://github.com/yangyanzhan/code-camp"
__blog__ = "https://yangyanzhan.github.io"

class Solution:
    def romanToInt(self, s: str) -> int:
        bases = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]
        romans = ["M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"]
        res = 0
        for i in range(len(bases)):
            base = bases[i]
            roman = romans[i]
            while s.startswith(roman):
                res += base
                s = s[len(roman):]
        return res
